File: Litext/LTXAttachment.m
//
//  LTXAttachment.m
//  Litext
//
//  Created by Cyandev on 2022/5/9.
//

#import <CoreText/CoreText.h>

#import "LTXAttachment.h"

NSAttributedStringKey LTXAttachmentAttributeName = @"LTXAttachment";

static void _LTXRunDelegateDeallocCallback(void *refCon) {
    // Stub.
}

static double _LTXRunDelegateGetAscentCallback(void *refCon) {
    LTXAttachment *attachment = (__bridge id) refCon;
    return attachment.size.height * 0.9;
}

static double _LTXRunDelegateGetDescentCallback(void *refCon) {
    LTXAttachment *attachment = (__bridge id) refCon;
    return attachment.size.height * 0.1;
}

static double _LTXRunDelegateGetWidthCallback(void *refCon) {
    LTXAttachment *attachment = (__bridge id) refCon;
    return attachment.size.width;
}

@implementation LTXAttachment {
    CTRunDelegateRef _ctRunDelegate;
}

- (void)dealloc {
    if (_ctRunDelegate) {
        CFRelease(_ctRunDelegate);
        _ctRunDelegate = nil;
    }
}

- (id)runDelegate {
    if (!_ctRunDelegate) {
        CTRunDelegateCallbacks callbacks = { kCTRunDelegateVersion1, NULL, NULL, NULL, NULL };
        callbacks.dealloc = &_LTXRunDelegateDeallocCallback;
        callbacks.getAscent = &_LTXRunDelegateGetAscentCallback;
        callbacks.getDescent = &_LTXRunDelegateGetDescentCallback;
        callbacks.getWidth = &_LTXRunDelegateGetWidthCallback;
        // TODO: use a weak proxy instead of passing an unsafe unretained opaque pointer.
        _ctRunDelegate = CTRunDelegateCreate(&callbacks, (__bridge void *) self);
    }
    return (__bridge id) _ctRunDelegate;
}

@end

File: Sources/Litext/LTXLabel/Attachments/LTXAttachment.swift
//
//  Created by Lakr233 & Helixform on 2025/2/18.
//  Copyright (c) 2025 Litext Team. All rights reserved.
//

import CoreText
import Foundation

#if os(watchOS)
    import SwiftUI
#endif

@MainActor
open class LTXAttachment {
    static let descentFraction: CGFloat = 0.1

    open var size: CGSize

    #if !os(watchOS)
        /// The platform view to embed as an inline attachment (iOS/macOS/tvOS/visionOS).
        open var view: LTXPlatformView?
    #else
        /// The SwiftUI view to embed as an inline attachment (watchOS).
        open var swiftUIView: AnyView?
    #endif

    public init() {
        size = .zero
    }

    open func attributedStringRepresentation() -> NSAttributedString {
        #if !os(watchOS)
            if let view = view as? LTXAttributeStringRepresentable {
                return view.attributedStringRepresentation()
            }
        #endif
        return NSAttributedString(string: " ")
    }

    open var runDelegate: CTRunDelegate {
        var callbacks = CTRunDelegateCallbacks(
            version: kCTRunDelegateVersion1,
            dealloc: { refCon in
                Unmanaged<LTXAttachment>.fromOpaque(refCon).release()
            },
            getAscent: { refCon in
                let attachment = Unmanaged<LTXAttachment>.fromOpaque(refCon).takeUnretainedValue()
                return attachment.size.height * (1 - LTXAttachment.descentFraction)
            },
            getDescent: { refCon in
                let attachment = Unmanaged<LTXAttachment>.fromOpaque(refCon).takeUnretainedValue()
                return attachment.size.height * LTXAttachment.descentFraction
            },
            getWidth: { refCon in
                let attachment = Unmanaged<LTXAttachment>.fromOpaque(refCon).takeUnretainedValue()
                return attachment.size.width
            }
        )

        let unmanagedSelf = Unmanaged.passRetained(self)
        guard let delegate = CTRunDelegateCreate(&callbacks, unmanagedSelf.toOpaque()) else {
            unmanagedSelf.release()
            fatalError("Unable to create CTRunDelegate for LTXAttachment")
        }
        return delegate
    }
}
